### ISSUE TITLE
code vector vs target embedding

### ISSUE THREAD

**Issue by DavidSUN2** [tech-srl/code2vec#56]
Hi Uri,
I'm confused about code vecor and target embedding. In Readme:

```
Exporting the trained target (method name) embeddings:
python3 code2vec.py --load models/java14_model/saved_model_iter8 --save_t2v models/java14_model/targets.txt
```
```
Exporting the code vectors for the given code examples
The flag --export_code_vectors allows to export the code vectors for the given examples.
```
1. To my understanding, code vectors are generated after combined context vectors and attention weights, target embeddings are weights between code vector layer and softmax layer. So, they are different. Each method name can be represented by code vector or target embedding (if the method name is in the target/tag dictionary). Is my understanding right?
2. When exporting the code vectors for the given code examples, the output file doesn't contain method names. Is it possible to add method named in the output file?
3. When using the gensim python package, in your example, you use target embeddings. Can we use code vectors as well?
Thanks.

**Comment by urialon** [comment]
Hi @DavidSUN2 ,
1. The code vectors and the target embeddings are trained to be close to each other, but they are different as you say, this is right. The vector of a specific method name (e.g., `sort`) is shared among all the methods that are labeled as `sort`, whereas the code vector of each of these examples is slightly different than the others.
2. Yes, it is possible to add method names, although sometimes you need the code vector of a new given method and you don't have any given method name. I don't have the capacity to do it myself right now, but it is not supposed to be difficult at all. If you're trying to do that and have any questions feel free to ask.
3. Definitely yes, as long as you have a given label (=method name or just ID number) for each code vector.

I hope it helps. Feel free to ask any more questions.

Best,
Uri

**Comment by DavidSUN2** [comment]
Hi @urialon,
Thanks for your reply.
`The code vectors and the target embeddings are trained to be close to each other`
I'm confused about this. Why are they close to each other by training? Could you please do more explanation? Thanks.
In addition, target embeddings are weights between code vector layer and softmax layer. Is this right?

**Comment by urialon** [comment]
Hi @DavidSUN2,
Sorry, it is indeed **not** trivial - 

First, yes, target embeddings are weights vectors between the code vector and the softmax layer.

Second, during training, the softmax (+ cross-entropy loss) encourages the code vector to have a large value of dot-product with the "correct" target embedding (the target embedding of the true label), and a low dot-product with each of the rest of the target embeddings. So eventually, it makes the code vector its corresponding true-target embedding be close to each other in the euclidean space.

This is a characteristic of softmax+cross entropy which is not specific to code2vec.

**Comment by DavidSUN2** [comment]
Thanks a lot@urialon 
I understand.
